Add ITime (#149)
CliMA/ClimaUtilities.jl: docs/src/timemanager.md

fixed duration, such as `Week`, but unlike `Month`), for example
```@example example1
time1 = ITime(5; period = Hour(20), epoch = DateTime(2012, 12, 21))
```
All these quantities are accessible with functions:
```@example example1
@show "time1" counter(time1) period(time1) epoch(time1) date(time1)
```

Now that we know how to create `ITime`s, we can manipulate them. `ITime`s
support most (but not all) arithmetic operations.

For `ITime`s with a `epoch`, it is only possible to combine `ITime`s that
have the same `epoch`. `ITime`s can be composed to other times and
arithmetic operations propagate `epoch`s.

This works:
```@example example1
time1 = ITime(5; epoch = DateTime(2012, 12, 21))
time2 = ITime(15; epoch = DateTime(2012, 12, 21))
time1 + time2
```
This works too
```@example example1
time1 = ITime(5; epoch = DateTime(2012, 12, 21))
time2 = ITime(15)
time1 + time2
```
This does not work because the two `ITimes` don't have the same `epoch`
```@example example1
time1 = ITime(5; epoch = DateTime(2012, 12, 21))
time2 = ITime(15; epoch = DateTime(2025, 12, 21))
try  #hide
time1 + time2
catch err; showerror(stderr, err); end  #hide
```

> When dealing with binary operations, it is convenient to think of `ITimes` as
> dimensional quantities (quantities with units). In the previous examples,
> addition returns a new `ITimes`. Division will behave differently 

Reviewer: ```suggestion
When dealing with binary operations, it is convenient to think of `ITime`s as
dimensional quantities (quantities with units). In the previous examples,
addition returns a new `ITime`s. Division will behave differently 
```

are transformed to the 
```@example example1
time1 = ITime(5; period = Hour(1))
time2 = ITime(1; period = Minute(25))
time1 - time2
```
As we can see, the return value of 5 hours - 25 minutes is 275 minutes and it is
represented as 55 clock cycles of a period of 5 minutes. `Minute(5)` was picked
because it the greatest common divisor between `Hour(1)` and `Minute(25)`.

At any point, we can obtain a date or a number of seconds from an `ITime`:
```@example example1
time1 = ITime(5; period = Day(10), epoch = DateTime(2012, 12, 21))
date(time1)
```
And
```@example example1
time1 = ITime(5; period = Day(10), epoch = DateTime(2012, 12, 21))
seconds(time1)
```
In this, note that the `seconds` function always returns a Float64.

In this section, we saw that `ITime`s can be used to represent dates and times
and manipulated in a natural way.

`ITime`s support another feature, fractional times, useful for further
partitioning an interval of time.

### Dealing with times that cannot be represented

Sometimes, one need to work with fractions of a `period`. The primary
application of this is timestepping loops, which are typically divided in stages
which are a fraction of a timestep.

> In these cases, we round to the nearest amount representable as an ITime.
> Furthermore, these cases are constrained to only when an ITime is multipled by a
> float between zero and one. These are the only cases we consider, because the only
> instance in which the problem of handling fraction of a `period` is in the
> timestepping loops.

Reviewer: ```suggestion
In these cases, we round to the nearest amount representable as an `ITime`.
Furthermore, these cases are constrained to only when an `ITime` is multipled by a
float between zero and one. These are the only cases we consider, because the only
instance in which the problem of handling fraction of a `period` is in the
timestepping loops.
```

See the examples below.
```@repl example1
time = ITime(1; period = Hour(1), epoch = DateTime(2012, 12, 21))
0.3 * time # round down
0.5 * time # round down
0.7 * time # round up
1.2 * time # error because 1.2 > 1.0
-0.2 * time # error because -0.2 < 0.0
```

> If multiplication between an ITime and a float is needed, you most likely want
> to cast the ITime into a float. For example, in functions that compute
> tendencies, there are no operations that are of the form `t + a * dt` where `a`
> is a float, so it is safe to cast the ITime into a float as it will not lead to
> a loss of accuracy in time.

Reviewer: ```suggestion
If multiplication between an `ITime` and a float is needed, you most likely want
to cast the `ITime` into a float. For example, in functions that compute
tendencies, there are no operations that are of the form `t + a * dt` where `a`
is a float, so it is safe to cast the `ITime` into a float as it will not lead to
a loss of accuracy in time.
```

# TimeManager

`TimeManager` defines `ITime`, the time type for CliMA simulations, alongside
with various functions to work with it.

## `ITime`

`ITime` is a type to describe times and dates. The "I" in `ITime` stands for
_integer_: internally, `ITime` uses integers to represent times and dates,
meaning that operations with `ITime` are exact and do not occur in
floating-point errors.

`ITime` can be thought a combination of three quantities, a `counter`, a

> `period`, and (optionally) a `epoch`, with the `counter` counting how many

Reviewer: ```suggestion
`period`, and (optionally) an `epoch`, with the `counter` counting how many
```
